File: Library/Models/Song.cs
using Library.Enums;
using Library.Providers;
using Library.Utils;

namespace Library.Models;

public class Song : Music
{
    public List<User> Users { get; } = new();

    private readonly string artistIds;
    private List<Artist>? artists;

    public int Duration { get; }

    public Song(string id, string artistIds, int duration)
    {
        this.id = id; 
        this.artistIds = artistIds;
        Duration = duration;
    }

    public List<Artist> Artists
    {
        get
        {
            if (artists == null)
            {
                artists = new();
                foreach (var artistId in artistIds.Split(splitCharacter))
                {
                    foreach (var artist in Provider.Artists)
                    {
                        if (artist.Id != null)
                        {
                            if (artist.Id.Equals(artistId))
                            {
                                artists.Add(artist); break;
                            }
                        }
                    }
                }
            }
            return artists;
        }
    }

    public string ToString(Language language)
        => GetName(language) + " - " + Artists.GetNames(language, " & ");
}

File: Music.Short.Android/app/src/main/java/com/zitherharpmusic/zhmshort/model/Song.java
package com.zitherharpmusic.zhmshort.model;

import java.util.ArrayList;
import java.util.List;

public class Song extends Music {
    private final String artistIds;

    private List<Video> videos;
    private List<Artist> artists;

    public Song(String id, String artistIds) {
        this.id = id;
        this.artistIds = artistIds;
    }

    public List<Video> getVideos() {
        if (videos == null) {
            videos = new ArrayList<>();
            for (Video video : MusicProvider.getVideos()) {
                for (Song song : video.getSongs()) {
                    if (song.getId().equals(id)) {
                        videos.add(video);
                    }
                }
            }
        }
        return videos;
    }

    public List<Artist> getArtists() {
        if (artists == null) {
            artists = new ArrayList<>();
            for (Artist artist : MusicProvider.getArtists()) {
                for (String artistId : artistIds.split(SPLIT_CHARACTER)) {
                    if (artist.getId().equals(artistId)) {
                        artists.add(artist);
                    }
                }
            }
        }
        return artists;
    }

    public String toString(Language language) {
        return getName(language) + " - " + MusicUtils.getNames(getArtists(), language);
    }
}
